File: DiagonalMatrixTraversal.py
# Diagonal traverse of matrix

class Solution(object):
    def findDiagonalOrder(self, mat):
        """
        :type mat: List[List[int]]
        :rtype: List[int]
        """

        #logic to go up

        m = len(mat)    #Rows
        n = len(mat[0]) #Columns
        dir = True      #Boolean flag to track direction
        result=[]*(m*n) #Array of size m*n
        row = column = 0

        for i in range(m*n):
            result.append(mat[row][column])
            #logic to go up
            if dir:
                if column == n-1:
                    row+=1
                    dir=False
                elif row == 0:
                    column+=1
                    dir=False
                else:
                    row-=1
                    column+=1
            #logic to go down
            else: 
                if row == m-1:
                    column+=1
                    dir=True
                elif column == 0:
                    row+=1
                    dir=True
                else:
                    row+=1
                    column-=1
        return result


        
obj = Solution()
mat = [[1,2,3],[4,5,6],[7,8,9]]
result = obj.findDiagonalOrder(mat)
print(result)

File: DiagonalMatrixTraversal.java

/**

 Approach: Matrix traversal diagonally

 Working:
     for each iteration
         traverse diagonally from bottom to top direction 1
         or traverse diagonally from top to bottom direction -1


 Time Complexity: O(m*n)
 Space Complexity: O(1) // since the method requires to return an 1d of matrix elements, its constat
 */

class DiagonalMatrixTraversal {
    public int[] findDiagonalOrder(int[][] mat) {

        int m = mat.length;
        int n = mat[0].length;
        int[] result = new int[m*n];
        int row = 0 ,col = 0, index = 0;
        int dir =1;

        while(index<m*n) {
            result[index] = mat[row][col];
            index++;

            //bottom to top
            if(dir == 1) {

                //col max index limit reached
                if(col == n-1) {
                    dir = -1;
                    row++;
                }
                else if(row == 0) { // row min index limit reached
                    dir = -1;
                    col++;
                }
                else {
                    row--;
                    col++;
                }
            }
            else {
                //direction -1
                //top to bottom
                if(row == m-1) {
                    dir = 1;
                    col++;
                }
                else if(col == 0){
                    dir = 1;
                    row++;
                }
                else{
                    row++;
                    col--;
                }
            }

        }
        return result;

    }
}
